lib/Beam/Emitter.pm

```diff
@@ -15,20 +15,27 @@ use Beam::Listener;
 # The event listeners on this object, a hashref of arrayrefs of
 # EVENT_NAME => [ Beam::Listener object, ... ]
 
-=attr listeners
-
-An array containing the listeners which have subscribed to this
-emitter.  The array elements are instances of classes which inherit
-from L<Beam::Listener>
-
-=cut
-
-has listeners => (
+has _listeners => (
     is      => 'ro',
     isa     => HashRef,
     default => sub { {} },
 );
 
+=method listeners( event_name )
+
+Returns an array containing the listeners which have subscribed to the
+specified event from this emitter.  The array elements are instances
+of sub-classes of L<Beam::Listener>.
+
+=cut
+
+sub listeners {
+
+    my ( $self, $name ) = @_;
+
+    return $self->_listeners->{$name} || [];
```

Comment:
Could you deref this arrayref and return the list instead? Returning the arrayref from `listeners()` could be a potential source of spooky bugs: It could cause unrelated code to think it owned that ref and could do what it wanted (`shift`, `splice`, etc...). If that ref gets passed around too far, the idea of who owns that ref gets lost. It's similar to C/C++ pointer ownership: The location of the error message (exception in Perl, segfault in C/C++) gets so far removed from the actual source of the problem that it becomes extremely difficult to debug. It's not so much that it's more or less useful to return a list or an arrayref, but it's protection against a rare corner case that leads to frustration.

It'd probably be easiest to deref the array and just return the list of listeners. Then people can make it into an array or an arrayref if they feel. In general I prefer returning lists to refs (both array and hash) mostly for that reason: Ownership is hard to know and even harder to fix.

If the emitter itself wants to edit that ref, it has the private `_listeners` hash to do so (though we should probably discourage that so that the data structure isn't part of the API). And if someone else wants to remove a listener, they could pass the listener into `unsubscribe`. In fact, the changes in this PR could enable us in the future to make that even easier to do...